From: Ubuntu Desktop bug work
Date: Fri, 14 Aug 2026 00:00:00 +0000
Subject: power: do not crash lid/suspend paths on missing proxies

external_monitor_is_connected() dereferenced a NULL DisplayConfig
proxy and a NULL GError on lid decisions. Inhibit and suspend
finish handlers used error->message when error was NULL. Lid-close
lock called ScreenSaver on a NULL proxy.

Forwarded: no
---
--- a/plugins/power/gpm-common.c	2026-03-15 00:42:54.000000000 +0000
+++ b/plugins/power/gpm-common.c	2026-08-14 09:06:11.167861278 +0000
@@ -290,10 +290,15 @@
         if (get_mock_external_monitor_file ())
                 return mock_external_monitor_is_connected ();
 
+        if (!display_config)
+                return FALSE;
+
         /* This needs to be a synchronous call, an up to date state is needed
          * in response to lid changes. */
 
         connection = g_dbus_proxy_get_connection (G_DBUS_PROXY (display_config));
+        if (!connection)
+                return FALSE;
         variant = g_dbus_connection_call_sync (connection,
                                                "org.gnome.Mutter.DisplayConfig",
                                                "/org/gnome/Mutter/DisplayConfig",
@@ -308,7 +313,8 @@
                                                NULL,
                                                &error);
         if (!variant) {
-                g_debug ("Failed to get property 'HasExternalMonitor': %s", error->message);
+                g_debug ("Failed to get property 'HasExternalMonitor': %s",
+                         error ? error->message : "unknown error");
                 return FALSE;
         }
 
--- a/plugins/power/gsd-power-manager.c	2026-03-15 00:42:54.000000000 +0000
+++ b/plugins/power/gsd-power-manager.c	2026-08-14 09:06:11.715861599 +0000
@@ -1157,7 +1157,7 @@
                                            res,
                                            &error);
         if (result == NULL)
-                g_warning ("%s: %s", msg, error->message);
+                g_warning ("%s: %s", msg, error ? error->message : "unknown error");
 }
 
 static void
@@ -1635,6 +1635,9 @@
 {
         gboolean do_lock;
 
+        if (!manager->screensaver_proxy)
+                return;
+
         do_lock = g_settings_get_boolean (manager->settings_screensaver,
                                           "lock-enabled");
         if (!do_lock) {
@@ -2754,8 +2757,10 @@
 
         res = g_dbus_proxy_call_with_unix_fd_list_finish (proxy, &fd_list, result, &error);
         if (res == NULL) {
-                g_warning ("Unable to inhibit lid switch: %s", error->message);
-                g_error_free (error);
+                g_warning ("Unable to inhibit lid switch: %s",
+                           error ? error->message : "unknown error");
+                manager->inhibit_lid_switch_taken = FALSE;
+                g_clear_error (&error);
         } else {
                 g_variant_get (res, "(h)", &idx);
                 manager->inhibit_lid_switch_fd = g_unix_fd_list_get (fd_list, idx, &error);
@@ -2824,8 +2829,10 @@
 
         res = g_dbus_proxy_call_with_unix_fd_list_finish (proxy, &fd_list, result, &error);
         if (res == NULL) {
-                g_warning ("Unable to inhibit suspend: %s", error->message);
-                g_error_free (error);
+                g_warning ("Unable to inhibit suspend: %s",
+                           error ? error->message : "unknown error");
+                manager->inhibit_suspend_taken = FALSE;
+                g_clear_error (&error);
         } else {
                 g_variant_get (res, "(h)", &idx);
                 manager->inhibit_suspend_fd = g_unix_fd_list_get (fd_list, idx, &error);
